llvm/lib/SYCLLowerIR/LowerESIMDVecArg.cpp

```diff
@@ -0,0 +1,432 @@
+  for (auto &F : M) {
+    for (auto BB = F.begin(), BBEnd = F.end(); BB != BBEnd; ++BB) {
+      DenseMap<Instruction *, Instruction *> OldNewInst;
+      for (auto OI = BB->begin(), OE = BB->end(); OI != OE; ++OI) {
+        SmallVector<Value *, 6> Operands;
+        bool HasGlobals = false;
+        auto &Inst = (*OI);
+        for (unsigned int OP = 0, OPE = Inst.getNumOperands(); OP != OPE;
+             ++OP) {
+          auto opnd = Inst.getOperand(OP);
+          if (!isa<ConstantExpr>(opnd)) {
+            Operands.push_back(opnd);
+            continue;
+          }
+          GlobalVariable *OldGlobal = nullptr;
+          auto OldGlobalVar = hasGlobalConstExpr(opnd, OldGlobal);
+          if (OldGlobalVar &&
+              OldNewGlobal.find(OldGlobal) != OldNewGlobal.end()) {
+            HasGlobals = true;
+            auto NewGlobal = OldNewGlobal[OldGlobal];
+            assert(NewGlobal && "Didnt find new global");
+            Operands.push_back(
+                createNewConstantExpr(NewGlobal, OldGlobal->getType(), opnd));
+          } else {
+            Operands.push_back(opnd);
+          }
+        }
+
+        if (HasGlobals) {
+          Instruction *NewInst = nullptr;
+          if (isa<CallInst>(&Inst)) {
+            assert(isa<CallInst>(&Inst) && "Expecting call instruction");
+            // pop last parameter which is function declaration
+            auto CallI = cast<CallInst>(&Inst);
+            Operands.pop_back();
+            NewInst =
+                CallInst::Create(CallI->getFunctionType(),
+                                 CallI->getCalledFunction(), Operands, "");
+            cast<CallInst>(NewInst)->setTailCallKind(CallI->getTailCallKind());
+          } else if (isa<StoreInst>(&Inst)) {
+            auto StoreI = cast<StoreInst>(&Inst);
+            NewInst = new StoreInst(Operands[0], Operands[1],
+                                    StoreI->isVolatile(), StoreI->getAlign());
+          } else if (isa<LoadInst>(&Inst)) {
+            auto LoadI = cast<LoadInst>(&Inst);
+            NewInst =
+                new LoadInst(Inst.getType(), Operands[0], LoadI->getName(),
+                             LoadI->isVolatile(), LoadI->getAlign());
+          } else {
+            assert(false && "Not expecting this instruction with global");
+          }
+          OldNewInst[&Inst] = NewInst;
+          NewInst->copyMetadata(Inst);
+        }
+      }
+
+      for (auto Replace : OldNewInst) {
+        ReplaceInstWithInst(Replace.first, Replace.second);
+      }
+    }
+  }
+}
+
+// This function creates new global variables of type vector* type
+// when old one is of simd* type.
+void ESIMDLowerVecArgPass::fixGlobals(Module &M) {
+  for (auto &G : M.getGlobalList()) {
+    auto NewTy = argIsSimdPtr(&G);
+    if (NewTy && !G.user_empty()) {
+      // Peel off ptr type that argIsSimdPtr applies
+      NewTy = NewTy->getPointerElementType();
+      auto ZeroInit = new APInt(32, 0);
```

Comment:
You're assuming that the element type is i32, I think. The rest of the code has a general approach to the contained vector type. I think ConstantAggregateZero::get(NewTy) is what you want here.

Shouldn't you also be checking to see what kind of initializer was used for the original variable?

Reply:
fixed.

i dont think static initialization is supported for on device code.